data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependantTokenProviderBase.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class CloudDependantTokenProviderBase extends TokenProviderBase {
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL =
+            "Error determining scope due to invalid Kusto Service Resource URL";
+    protected final Set<String> scopes = new HashSet<>();
+    protected CloudInfo cloudInfo = null;
+
+    CloudDependantTokenProviderBase(@NotNull String clusterUrl) throws URISyntaxException {
+        super(clusterUrl);
+    }
+
+    protected void onCloudInfoInitialized() throws DataClientException, DataServiceException {
```

Comment:
nit: "on" doesn't sound intuitive to me. I think "after" makes more sense.

Reply:
Renamed to `initializeWithCloudInfo` now that it's a parameter. Is that ok?

Reply:
Perfect!